README.md
# Sudoku16x16Explainer

Sudoku16x16Explainer is my modifications to SudokuExplainer to solve 16x16 sudokus.

## Puzzle Format

- The GUI displays solved cells and candidates using numbers 1-16.
- Load/Save and Copy/Paste use A-P for the numbers 1-16 - The other formats: 0-9A-F, 1-9A-G, 1-16 are supported via 'Options', select the format BEFORE Load/Save and Copy/Paste operations.
- Candidate input via keyboard is through keys A-P/a-p for the numbers 1-16, and keys 1-9 for numbers 1-9 - it is a better experience to select candidates using the mouse (left-click to set a candidate, right-click to remove a candidate).

A puzzle converter to the A-P format used by Sudoku16Explainer is [here](https://github.com/1to9only/convert16x16).

## Usage - GUI
```
java.exe -jar Sudoku16Explainer.jar
```
![](/images/sample1.png)

## Usage - serate
```
java.exe -Xrs -Xmx1g -cp Sudoku16Explainer.jar diuf.sudoku.test.serate --input=puzzle(s).txt --output=output.txt
```
## Usage - hints, pencilmarks
```
java.exe -Xrs -Xmx1g -cp Sudoku16Explainer.jar diuf.sudoku.test.hints --input=puzzle(s).txt

java.exe -Xrs -Xmx1g -cp Sudoku16Explainer.jar diuf.sudoku.test.hints --input=puzzle(s).txt 1>output.txt
java.exe -Xrs -Xmx1g -cp Sudoku16Explainer.jar diuf.sudoku.test.hints --input=puzzle(s).txt 1>output.txt 2>nul
```
STDOUT output (long hints) should (always) be redirected to a file.\
STDERR output (short hints) is for debug, it shows progress on screen, if not needed redirect this to a file or to nul.

## Usage - analyze (Tester)
```
java.exe -Xrs -Xmx1g -cp Sudoku16Explainer.jar diuf.sudoku.test.Tester puzzles.txt puzzles.log
```
## Usage - manual
```
java.exe -cp Sudoku16Explainer.jar diuf.sudoku.test.serate -m
```
## Testing

The program is functional, and it had limited testing. The first two puzzles it was tested with (and solved) are below:

This puzzle was posted by monstersudoku [here](http://forum.enjoysudoku.com/post6769.html#p6769):
```
.F..G..O.P.M.A..
L..ACE...B.DF.O.
.O...N.BF.......
..JP..M...C..EBI
.AC.PDO..JH...E.
GMEN......B.OC..
...F....LI..HNP.
P.....N.D.K....J
A....O.M.L.....P
.CNM..HD....I...
..PD.I......CHLK
.I...CK..DPN.BJ.
NKL..M...O..BJ..
.......CI.M...N.
.J.OI.F...LKP..G
..M.O.L.H..J..C. ED=2.3/1.2/1.2

Analyzing Sudoku #1
145 Hidden Single
3 Direct Pointing
4 Direct Hidden Pair
2 Naked Single
Hardest technique: Naked Single
Difficulty: 2.3
```

This is monstersudoku's puzzle modified by m_b_metcalf from [here](http://forum.enjoysudoku.com/post258131.html#p258131):
```
.E..F..N.O.L.P..
K..PBD...A.CE.N.
.....M.AE.......
..IO..L...B..D.H
.PB.OCN..IG...D.
FLDM........NB..
...E....KH..GMO.
......M.C.J....I
P....N.L.K.....O
.BML..GC....H...
..OC.H......BGKJ
.H...BJ...OM.AI.
MJK......N..AI..
.......BH.L...M.
.I..H......JO..F
..L...K.G.....B. ED=9.2/1.2/1.2

Analyzing Sudoku #1
152 Hidden Single
2 Direct Pointing
5 Direct Hidden Pair
1 Naked Single
1 Direct Hidden Triplet
22 Pointing
7 Claiming
2 Naked Pair
1 X-Wing
2 Hidden Pair
1 Naked Triplet
2 Hidden Triplet
1 Jellyfish
4 Turbot Fish
18 Forcing Chain
1 Bidirectional Cycle
3 Region Forcing Chains
6 Dynamic Region Forcing Chains
5 Dynamic Cell Forcing Chains
1 Dynamic Contradiction Forcing Chains
1 Dynamic Double Forcing Chains
Hardest technique: Dynamic Region Forcing Chains
Difficulty: 9.2
```

Sudoku16Explainer can serate:
- tarek's 206 tough puzzles<sup>1</sup> from [here](http://forum.enjoysudoku.com/post275765.html#p275765) (range: ED=3.0/1.2/1.2 - ED=9.7/1.2/1.2)
- tarek-first20 puzzles from [here](http://forum.enjoysudoku.com/post276187.html#p276187) (range: ED=2.6/1.2/1.2 - ED=9.4/1.2/1.2)

## (Non-) Issues

- Sudoku16Explainer can take a long time to solve the hardest 16x16 sudokus - be patient ...

- Sudoku16Explainer can generate a puzzle, the Difficulty level is set to Diabolical<sup>2</sup> at Maximum difficulty and cannot be changed ... - it takes some time to generate a puzzle ... - be patient ...

This is an example puzzle generated by Sudoku16Explainer:
```
....GF......M...
.D.F.L.A.O..KGI.
.CK.H.JE.M.P.F..
BN.PIDK.G...H.J.
..F.J.LKM..HDP.B
.......O....G.LN
.PI...HG.L.AFC..
...ND....CFO.HE.
.BG.EAP....IN...
..NDK.I.OB...AP.
PM.I....C.......
O.JEL..HFP.N.D..
.F.G...J.KHLE.CA
..H.A.E.PF.C.MN.
.OMB..D.E.G.P.F.
...K......OM.... ED=7.4/1.2/1.2

Analyzing Sudoku #1
131 Hidden Single
3 Direct Hidden Pair
6 Naked Single
1 Direct Hidden Triplet
12 Pointing
5 Claiming
2 Naked Pair
5 Hidden Pair
2 Naked Triplet
2 Swordfish
2 Hidden Triplet
1 XY-Wing
2 Direct Hidden Quad
1 Naked Quintuplet
1 Naked Sextuplet
1 Turbot Fish
1 Bidirectional Cycle
15 Forcing Chain
Hardest technique: Forcing Chain
Difficulty: 7.4
```


<sup>1</sup> tarek's 206 tough puzzles are in file test4.1g from [here](https://github.com/1to9only/convert16x16), use convert16x16.exe to convert to A-P format!

<sup>2</sup> Sudoku16Explainer has generated puzzles ranging from ED=2.0/1.2/1.2 to ED=10.3/1.2/1.2.

.



Changes.txt
Changes:

20230405
- change to showing candidate masks

20220510
- several small updates

20220411
- various small changes

20220323
- add fixed lksudoku's fix to Chaining
- add step timings in hints output

20220316
- small code tidy up

20220307
- changes in Chaining.java, may affect some ratings!

20220209
- add Check uniqueness only

20220202
- re-enable some solving techniques
- re-enable thread priority in Solver

20220112
- add chains nodes in hints output

20211125
- change to hints output

20211116
- revert an earlier change

20211109
- detect possible corrupted config (json) file

20211019
- a few cosmetic changes

20210921
- hints improvements
- built using jdk-8u301

20210405
- a few cosmetic changes

20210316
- a few fixes, and a few improvements

20210118
- some small fixes, changes, improvements:
- Dark theme, 'Restart...' and 'Save as image...' options, 'Apply Singles' button

20200915
- fixed one old bug, and other small changes

20191118
- re-enabled commented out techniques, 8 solving techniques are disabled
- changed to PLAIN font for GUI grid, black given cells, blue solved cells

20191104
- disabled changing thread priority in Solver.
- made Full Symmetry only initial option in GenerateDialog.
- changed ratings Naked and Hidden: Quintuplets, Sextuplets, Septuplets (5.4) - septuplets are disabled.
- changed ratings Fishes: Starfishes, Whales, Leviathans (5.5).
- added support for 0-9A-F, 1-9A-G, 1-16 puzzle formats, in GUI only.
- added Naked Octuplet, Hidden Octuplet, Loch Ness Monster (5.6) - all are elusive, and disabled.
- slightly improved SolvingTechnique, NakedSingle, Chaining, Grid.
- fixed parallel processing threads finishing order

20191015
- modified to solve 16x16 sudokus: grid is A-P for given, . or 0 for empty cell
- disabled Unique Loops (UL), Aligned Pair Exclusion (APE) and AlignedTripletExclusion (ATE) - this is because of software limitations in Permutations.java.
- added Direct Hidden Quads, with rating of 4.3.
- added Naked Quintuplets, Sextuplets, Septuplets, all with same rating as for Naked Quad (5.0).
- added Fishes: Starfishes, Whales, Leviathans, all with same rating as for Jellyfish (5.2).
- added Hidden Quintuplets, Sextuplets, Septuplets, all with same rating as for Hidden Quad (5.4).
- added parallel processing, based on code from [SukakuExplainer](https://github.com/SudokuMonster/SukakuExplainer).
- added Difficulty total, based on code from [NewerSudokuExplainer](https://github.com/Sunnie-Shine/NewerSudokuExplainer).
- added hints.java

Source:
- 1.2.1 - Nicolas Juillerat's SudokuExplainer
- 1.2.1.3 - gsf's (Glenn Fowler) serate modifications from [here](http://gsf.cococlyde.org/download/sudoku/serate.tgz) used to rate sudoku puzzles in the [Patterns Game](http://forum.enjoysudoku.com/patterns-game-t6290.html).
- updated Reader code to handle most 9x9 grid formats: grid is 1-9 for given, . or 0 for empty cell
- added **Undo** based on code from [here](https://github.com/Itsukara/SudokuExplainerPlus)
- fixed mouse click issue in unsolved cell to select potential candidate
- added **Candidate Masks** based on code from [here](https://github.com/blindlf/SudokuExplainer)
- added Chaining.java changes from [here](http://forum.enjoysudoku.com/post280341.html)
- commented out all assert statements



diuf/sudoku/SolvingTechnique.java
/*
 * Project: Sudoku Explainer
 * Copyright (C) 2006-2007 Nicolas Juillerat
 * Available under the terms of the Lesser General Public License (LGPL)
 */
package diuf.sudoku;

public enum SolvingTechnique {

    HiddenSingle("Hidden Single 1.0-1.5"),
    DirectPointing("Direct Pointing 1.7"),
    DirectHiddenPair("Direct Hidden Pair 2.0"),
    NakedSingle("Naked Single 2.3"),
    DirectHiddenTriplet("Direct Hidden Triplet 2.5"),
    PointingClaiming("Pointing & Claiming 2.6-2.8"),
    NakedPair("Naked Pair 3.0"),
    XWing("X-Wing 3.2"),
    HiddenPair("Hidden Pair 3.4"),
    NakedTriplet("Naked Triplet 3.6"),
    Swordfish("Swordfish 3.8"),
    HiddenTriplet("Hidden Triplet 4.0"),
    XYWing("XY-Wing 4.2"),
    DirectHiddenQuad("Direct Hidden Quad 4.3"),
    XYZWing("XYZ-Wing 4.4"),
    UniqueLoop("Unique Rectangle / Loop 4.5+"),
    NakedQuad("Naked Quad 5.0"),
    Jellyfish("Jellyfish 5.2"),
    HiddenQuad("Hidden Quad 5.4"),

    NakedQuintuplet("Naked Quintuplet 5.4"),
    HiddenQuintuplet("Hidden Quintuplet 5.4"),
    NakedSextuplet("Naked Sextuplet 5.4"),
    HiddenSextuplet("Hidden Sextuplet 5.4"),
    NakedSeptuplet("Naked Septuplet 5.4"),
    HiddenSeptuplet("Hidden Septuplet 5.4"),

    Starfish("Starfish 5.5"),
    Whale("Whale 5.5"),
    Leviathan("Leviathan 5.5"),

    NakedOctuplet("Naked Octuplet 5.6"),
    HiddenOctuplet("Hidden Octuplet 5.6"),
    LochNessMonster("Loch Ness Monster 5.6"),

    BivalueUniversalGrave("Bivalue Universal Grave 5.6,5.7+"),
    AlignedPairExclusion("Aligned Pair Exclusion 6.2"),
    ForcingChainCycle("Forcing Chains & Cycles 6.6+,7.0+"),
    AlignedTripletExclusion("Aligned Triplet Exclusion 7.5"),
    NishioForcingChain("Nishio Forcing Chains 7.5+"),
    MultipleForcingChain("Multiple Forcing Chains 8.0+"),
    DynamicForcingChain("Dynamic Forcing Chains 8.5+"),
    DynamicForcingChainPlus("Dynamic Forcing Chains (+) 9.0+"),
    NestedForcingChain("Nested Forcing Chains 9.5+");

    private final String name;

    private SolvingTechnique(String name) {
        this.name = name;
    }

    @Override
    public String toString() {
        return name;
    }

}
